### DISCUSSION TITLE
flatmap versus tilemaker

### DISCUSSION THREAD

**Discussion by stephan75** [onthegomap/planetiler#16] [General]
Hello Michael  @msbarry 

recently I came across flatmap and this well documented github repo ... awesome efforts!!

Subsequently I added your project to my github watchlist where I have a lot of OSM related software etc.

So you mention https://github.com/systemed/tilemaker as a similar approach to produce OSM based MVT files.

Though I have not enough skills and hardware to run flatmap or tilemaker on my own, I ask myself:

What is the difference between tilemaker and flatmap?

So what features do they have in common, and where do they differ?

Of course I already read documentation and blogs for them both, but I am not able to put any answer in own words.

Maybe both projects can benefit from each other?

Or why are there some issues in one of them (mainly tilemaker) that seem to not occure in the other one?

(This text has been copied from a similar question in the tilemaker repo discussions section)

**Comment by msbarry** [comment]
Hello, thanks for reaching out!

Flatmap and tilemaker are similar in that they are both single executables with no external dependencies that read shapefiles and osm.pbf files and emit an mbtiles file containing mapbox vector tiles. They both let you define custom schemas with attributes that drop out below certain zoom levels and that merge adjacent polygons/linestrings below zoom levels.

They differ in that:
- Tilemaker's documentation says it's not meant for processing the entire planet, whereas I built flatmap specifically to generate tiles for the planet by offloading as much temporary storage to disk as possible (only node locations and relation membership ends up being stored in-memory). Tilemaker has been making progress in this area since I first played around with it, but it looks like it's taking [7-8 hours and 144GB RAM to generate the planet with tilemaker](https://github.com/systemed/tilemaker/discussions/336) vs. 1-1.5 hours and 100GB RAM for flatmap on equivalent hardware.
- Tilemaker has a convenient lua API for defining custom profiles, whereas with flatmap you need to write it in java code.
- Tilemaker's openmaptiles schema is close to the openmaptiles project, whereas flatmap's "basemap" profile is an exact port of openmaptiles v3.12.2 (although it may drift over time as openmaptiles evolves).
- Flatmap is written in Java - I use [graphhopper](https://github.com/graphhopper/graphhopper) for routing so that will eventually let me share code between the two and style roads based on whether they are routable or not.  Tilemaker is written in C++/lua so you might get that same benefit if you use [OSRM](https://github.com/Project-OSRM/osrm-backend) for routing.
- Flatmap lets you style ways based on the routes they are a part of, but I don't think tilemaker supports that yet (I imagine it wouldn't be too hard to add though).

> Or why are there some issues in one of them (mainly tilemaker) that seem to not occure in the other one?

Tilemaker has been around for longer and lots of people have used it, no one has found the issues in flatmap yet 😉

**Reply by stephan75** [reply]
Many thanks for your detailed information!
Thus I can better imagine for what use cases each software can be a better solution.

I am asking myself whether an own article page for flatmap could be created in the **OSM wiki** ... I didn't find it there so far, right?

Some points of your listing above could be documented there, too.

And there sould be a decision before whether you will keep the flatmap name or not, like mentioned in #15

**Reply by stephan75** [reply]
Okay, when the name change is done now, let's see whether I or someone else have some moments to create an item in the OSM wiki about planetiler ...

**Reply by msbarry** [reply]
Thanks! Is there any section you thought it should be listed under or just a standalone page? I took a brief look but couldn't figure out what part of the wiki it might fit under.

**Reply by stephan75** [reply]
Well, I just created that wiki page now: see https://wiki.openstreetmap.org/wiki/Planetiler

But the wiki source code still needs some tweaking:

Be aware that I included the template:Software ... see https://wiki.openstreetmap.org/wiki/Template:Software

Many parameters need to be adjusted to planetiles features etc.

Maybe see the wiki sourcecode of https://wiki.openstreetmap.org/wiki/Tilemaker to get an idea what is important, and maybe what parameters can be deleted.

Further more, the actual wiki text which I copied from github can have some more internal wiki links like a target to https://wiki.openstreetmap.org/wiki/Vector_tiles ...

from there even backlinks to planetiler or tilemaker should be included.

@msbarry : do you already have an OSM wiki account? It is **not** automatically done when you have a normal OSM account.

**Reply by msbarry** [reply]
Thanks @stephan75! I added some more details to that page, and a link to tilemaker and planetiler under a new "Generators" section in https://wiki.openstreetmap.org/wiki/Vector_tiles.  Let me know what you think!
